src/hoi4_world/map/railways_converter.cpp

```diff
@@ -77,23 +79,78 @@ std::vector<std::pair<int, int>> ConvertVic3EndpointsToHoi4Endpoints(
          continue;
       }
 
-      hoi4_endpoints.emplace_back(hoi4_province_start_point, hoi4_province_end_point);
+      std::pair<int, int> pair{hoi4_province_start_point, hoi4_province_end_point};
+      hoi4_endpoints.emplace_back(pair);
+      handled_hoi4_endpoints.emplace(pair);
    }
 
    return hoi4_endpoints;
 }
 
 
-constexpr int urban_cost = 1;
-constexpr int plains_cost = 2;
-constexpr int forest_cost = 3;
-constexpr int hills_cost = 4;
-constexpr int desert_cost = 5;
-constexpr int marsh_cost = 6;
-constexpr int jungle_cost = 7;
-constexpr int mountain_cost = 8;
-constexpr int unhandled_cost = 100;
-int GetCostForTerrainType(std::string_view terrain_type)
+std::optional<float> GetInfrastructureLevel(int province, const hoi4::States& states)
+{
+   const auto state_itr = states.province_to_state_id_map.find(province);
+   if (state_itr == states.province_to_state_id_map.end())
+   {
+      return std::nullopt;
+   }
+
+   if (states.states.size() < state_itr->second)
+   {
+      return std::nullopt;
+   }
+
+   return states.states.at(state_itr->second - 1).GetVic3Infrastructure();
+}
+
+
+int DetermineRailwayLevel(int start_province, int end_province, const hoi4::States& states)
+{
+   const std::optional<float> start_infrastructure = GetInfrastructureLevel(start_province, states);
+   if (!start_infrastructure)
+   {
+      return 0;
+   }
+
+   const std::optional<float> end_infrastructure = GetInfrastructureLevel(end_province, states);
+   if (!end_infrastructure)
+   {
+      return 0;
+   }
+
+   const float total_infrastructure = *start_infrastructure + *end_infrastructure;
+   if (total_infrastructure > 3500.0F)
```

Comment:
Sure, just TODO a move of all these hardcodables to a config file.

Reply:
Long-term plan is to actually normalize them to the state of the world during conversions.